# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpConnectionBase.cpp
# Commit: 00bb17e2a6a7
# Full Hash: 00bb17e2a6a73f9c66e861adae426818ff4d8caa
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2023-09-21 21:20:56
# Regressor Bug: 1784261
# File Overlap Count: 1
# Description:
#   Bug 1784261 - Add telemetry to understand the reason of connection closure, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D188132
# ==============================================================================

diff -r e98350675c5a -r 00bb17e2a6a7 netwerk/protocol/http/HttpConnectionBase.cpp
--- a/netwerk/protocol/http/HttpConnectionBase.cpp	Thu Sep 21 08:42:26 2023 +0000
+++ b/netwerk/protocol/http/HttpConnectionBase.cpp	Thu Sep 21 08:59:03 2023 +0000
@@ -60,5 +60,42 @@
   Unused << mTrafficCategory.AppendElement(aCategory);
 }
 
+void HttpConnectionBase::ChangeConnectionState(ConnectionState aState) {
+  LOG(("HttpConnectionBase::ChangeConnectionState this=%p (%d->%d)", this,
+       static_cast<uint32_t>(mConnectionState), static_cast<uint32_t>(aState)));
+
+  // The state can't move backward.
+  if (aState <= mConnectionState) {
+    return;
+  }
+
+  mConnectionState = aState;
+}
+
+void HttpConnectionBase::RecordConnectionCloseTelemetry(nsresult aReason) {
+  /**
+   *
+   * The returned telemetry key has the format:
+   * "Version_EndToEndSSL_IsTrrServiceChannel_ExperienceState_ConnectionState"
+   *
+   * - Version: The HTTP version of the connection.
+   * - EndToEndSSL: Indicates whether SSL encryption is end-to-end.
+   * - IsTrrServiceChannel: Specifies if the connection is used to send TRR
+   *    requests.
+   * - ExperienceState: ConnectionExperienceState
+   * - ConnectionState: The connection state before closing.
+   */
+  auto key = nsPrintfCString("%d_%d_%d_%d_%d", static_cast<uint32_t>(Version()),
+                             mConnInfo->EndToEndSSL(),
+                             mConnInfo->GetIsTrrServiceChannel(),
+                             static_cast<uint32_t>(mExperienceState),
+                             static_cast<uint32_t>(mConnectionState));
+  SetCloseReason(ToCloseReason(aReason));
+  LOG(("RecordConnectionCloseTelemetry key=%s reason=%d\n", key.get(),
+       static_cast<uint32_t>(mCloseReason)));
+  Telemetry::Accumulate(Telemetry::HTTP_CONNECTION_CLOSE_REASON, key,
+                        static_cast<uint32_t>(mCloseReason));
+}
+
 }  // namespace net
 }  // namespace mozilla